modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/local/IgniteCacheLocalQueryDefaultTimeoutSelfTest.java

```diff
@@ -0,0 +1,153 @@
+    private static final long DEFAULT_QUERY_TIMEOUT = 1000;
+
+    /** */
+    private static final String QUERY = "select a._val, b._val from String a, String b";
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(igniteInstanceName);
+
+        CacheConfiguration<Integer, String> ccfg = new CacheConfiguration<>(DEFAULT_CACHE_NAME);
+        ccfg.setIndexedTypes(Integer.class, String.class);
+        ccfg.setCacheMode(LOCAL);
+
+        cfg.setCacheConfiguration(ccfg);
+        cfg.setDefaultQueryTimeout(DEFAULT_QUERY_TIMEOUT);
+
+        return cfg;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTestsStarted() throws Exception {
+        super.beforeTestsStarted();
+
+        startGrid(0);
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        super.afterTest();
+
+        G.ignite().cache(DEFAULT_CACHE_NAME).removeAll();
+    }
+
+    /**
+     * @param cache Cache.
+     */
+    private void loadCache(IgniteCache<Integer, String> cache) {
+        int p = 1;
+
+        for (int i = 1; i <= CACHE_SIZE; i++) {
+            char[] tmp = new char[256];
+            Arrays.fill(tmp, ' ');
+            cache.put(i, new String(tmp));
+
+            if (i / (float)CACHE_SIZE >= p / 10f) {
+                log().info("Loaded " + i + " of " + CACHE_SIZE);
+
+                p++;
+            }
+        }
+    }
+
+    /**
+     * Tests query execution with default query timeout.
+     */
+    @Test
+    public void testQueryZeroTimeout() {
+        testQuery(false, 1, TimeUnit.MILLISECONDS);
+    }
+
+    /**
+     * Tests query execution with query timeout.
+     */
+    @Test
+    public void testQueryTimeout() {
+        testQuery(true, 1, TimeUnit.SECONDS);
+    }
+
+    /**
+     * Tests cancellation.
+     */
+    private void testQuery(boolean timeout, int timeoutUnits, TimeUnit timeUnit) {
+        Ignite ignite = grid(0);
+
+        IgniteCache<Integer, String> cache = ignite.cache(DEFAULT_CACHE_NAME);
+
+        loadCache(cache);
+
+        SqlFieldsQuery qry = new SqlFieldsQuery(QUERY);
+
+        final QueryCursor<List<?>> cursor;
+        if (timeout) {
+            qry.setTimeout(timeoutUnits, timeUnit);
+
+            cursor = cache.query(qry);
+        }
+        else {
+            cursor = cache.query(qry);
+
+            ignite.scheduler().runLocal(new Runnable() {
+                @Override public void run() {
+                    cursor.close();
+                }
+            }, timeoutUnits, timeUnit);
+        }
+
+        try (QueryCursor<List<?>> ignored = cursor) {
+            cursor.iterator();
+
+            fail("Expecting timeout");
+        }
+        catch (Exception e) {
+            assertNotNull("Must throw correct exception", X.cause(e, QueryCancelledException.class));
```

Comment:
I do not  see how do we distinguish _manual cancel_ vs _explicit query timeout_ vs _default query timeout here_.